src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,401 @@
+        // If the function returns a promise we will wait for the promise to finish so we can
+        // serialize the result.
+        auto result = co_await callFuncAndSerialize(js, fn, serializedArgs);
+
+        auto builder = callContext.initResults(capnp::MessageSize { result.data.size() / 8 + 8, 0 });
+        if (result.caughtException) {
+          builder.initException().setV8Serialized(kj::mv(result.data));
+        } else {
+          builder.initValue().setV8Serialized(kj::mv(result.data));
         }
-    ));
+      }));
+    } catch(kj::Exception& e) {
+      if (auto desc = e.getDescription();
+          !jsg::isTunneledException(desc) && !jsg::isDoNotLogException(desc)) {
+        LOG_EXCEPTION("JsRpcTargetCall"_kj, e);
+      }
+    }
+    // Upon returning, we want to fulfill the callPromise so customEvent can continue executing.
+    KJ_DEFER(callFulfiller->fulfill(););
+    co_return;
+  }
+
+  KJ_DISALLOW_COPY_AND_MOVE(JsRpcTargetImpl);
+
+private:
+  // Prevents us from calling a function defined on the Object Prototype, or a registered method.
+  bool verifyCallable(jsg::Lock& js, v8::Local<v8::Object> workerObj, kj::StringPtr name) {
+    auto workerProto = workerObj->GetPrototype().As<v8::Object>();
+    auto objectProto = workerProto->GetPrototype().As<v8::Object>();
+
+    // Get() checks the Object and the prototype chain.
+    // This check tells us if the function we intend to call is the one on the Object prototype,
+    if (jsg::check(workerObj->Get(js.v8Context(), js.strExtern(name))) ==
+        jsg::check(objectProto->Get(js.v8Context(), js.strExtern(name)))) {
+      return false;
+    }
+    return isRegisteredName(name);
+  }
+
+  // The following names are registered by the Workers Runtime and cannot be called over RPC.
+  bool isRegisteredName(kj::StringPtr name) {
+    if (name == "fetch"_kjc ||
+        name == "connect"_kjc ||
+        name == "alarm"_kjc ||
+        name == "webSocketMessage"_kjc ||
+        name == "webSocketClose"_kjc ||
+        name == "webSocketError"_kjc) {
+      return true;
+    }
+    return false;
+  }
+
+  // If the `methodName` is a known public method, we'll return it.
+  inline kj::OneOf<v8::Local<v8::Function>, kj::String> tryGetFn(
+      Worker::Lock& lock,
+      kj::Maybe<kj::StringPtr> entrypointName,
+      IoContext& ctx,
+      kj::StringPtr methodName) {
+
+    auto& handler = KJ_REQUIRE_NONNULL(lock.getExportedHandler(entrypointName, ctx.getActor()),
+        "Failed to get handler to worker.");
+
+    auto handle = handler.self.getHandle(lock);
+    if (verifyCallable(lock, handle, methodName)) {
+      return kj::str("'", methodName, "' is a registered method and cannot be called over RPC."_kj);
+    }
+    auto methodStr = jsg::v8StrIntern(lock.getIsolate(), methodName);
+    auto fnHandle = jsg::check(handle->Get(lock.getContext(), methodStr));
+
+    // Get() doesn't handle private properties so we have to get this separately.
```

Comment:
[James confirmed](https://github.com/cloudflare/workerd/pull/1028#discussion_r1402004178) that `Get()` doesn't return private properties and that we have to use `GetPrivate()`, which we're doing below this line. I can rephrase this if it helps, but as I mentioned before, the client code cannot send requests for private methods because name interception doesn't allow it.

Reply:
> Are we doing that just in hopes of returning a better error message? I don't think that's really necessary.

Dropped here https://github.com/cloudflare/workerd/compare/a9254ed192e88344b38851274cb677fb36bd0fd6..9472684e8df98603e500ce62fe5021ba05cc50a5#diff-bbe685e15c44d4de0aed0b4b484f1c0fbe1427fccbf80ab9f65c6180f6632203L200-L211